src/queue.cpp

```diff
@@ -51,11 +58,91 @@ cvk_command_queue::~cvk_command_queue() {
     }
 }
 
-void cvk_command_queue::enqueue_command(cvk_command* cmd, _cl_event** event) {
+namespace {
+
+bool is_command_batchable(cvk_command* cmd) {
+    bool batchable_type = (cmd->type() == CL_COMMAND_NDRANGE_KERNEL) ||
+                          (cmd->type() == CL_COMMAND_TASK);
+    bool unresolved_user_event_dependencies = false;
+    bool unresolved_other_queue_dependencies = false;
+
+    for (auto ev : cmd->dependencies()) {
+        if (ev->is_user_event()) {
+            if (!ev->completed()) {
+                unresolved_user_event_dependencies = true;
+                break;
+            } else {
+                continue;
+            }
+        }
+
+        if ((ev->queue() != cmd->queue()) && !ev->completed()) {
+            unresolved_other_queue_dependencies = true;
+            break;
+        }
+    }
+
+    return batchable_type && !unresolved_user_event_dependencies &&
+           !unresolved_other_queue_dependencies;
+}
+
+} // namespace
+
+cl_int cvk_command_queue::enqueue_command(cvk_command* cmd, _cl_event** event) {
 
     std::lock_guard<std::mutex> lock(m_lock);
 
-    m_groups.back()->commands.push_back(cmd);
+    cl_int err;
+
+    bool is_kernel_command = cmd->type() == CL_COMMAND_NDRANGE_KERNEL ||
+                             cmd->type() == CL_COMMAND_TASK;
+    if (is_command_batchable(cmd)) {
+        if (!m_kernel_group) {
+            // Create a new kernel batch
+            m_kernel_group = new cvk_command_kernel_group(this);
+            m_groups.back()->commands.push_back(m_kernel_group);
+        }
+
+        // Add kernel to current batch
+        CVK_ASSERT(is_kernel_command);
+        err = m_kernel_group->add_kernel(static_cast<cvk_command_kernel*>(cmd));
+        if (err != CL_SUCCESS) {
+            return err;
+        }
+
+        // End kernel batch when size limit reached
+        if (m_kernel_group->batch_size() >= m_max_batch_size) {
+            if (!m_kernel_group->end()) {
+                return CL_OUT_OF_RESOURCES;
+            }
+            m_kernel_group = nullptr;
+        }
+    } else {
+        if (m_kernel_group) {
+            // End the current kernel batch
+            if (!m_kernel_group->end()) {
+                return CL_OUT_OF_RESOURCES;
+            }
+            m_kernel_group = nullptr;
+        }
+
+        if (is_kernel_command) {
+            // Create a standalone batch for a non-batchable kernel command
+            auto group = new cvk_command_kernel_group(this);
+            group->set_dependencies(cmd->dependencies());
+
+            err = group->add_kernel(static_cast<cvk_command_kernel*>(cmd));
+            if (err != CL_SUCCESS) {
+                return err;
+            }
+            if (!group->end()) {
+                return CL_OUT_OF_RESOURCES;
+            }
```

Comment:
Maybe this could be an init function to `cvk_command_kernel_group` that would prepare the group from a single command? Back to your question from the commit message:

>Having cvk_command_kernel::do_action() be a never-called method is a little awkward. [...]

Yeah and this forces us to create a group for every non-batchable kernel command but it's not the end of the world. I'm wondering whether the alternative wouldn't look a bit better though (not sure at all TBH).

Reply:
I've gone back to having a separate `cvk_command_kernel::build()` for non-batched kernels that creates its own command buffer, rather than having to create a standalone `cvk_command_kernel_group`. It's actually not as awkward as it was last time I tried it.